File: 2015/09/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";
import permutations from "../../../commons/javascript/permutations.mjs";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const REGEX = /^(.+) to (.+) = (\d+)$/

// --- Variables ---
let places = new Set();
let distances = {};
let min = Number.MAX_VALUE;
let max = Number.MIN_VALUE;

// --- Read and parse the input file ---
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	const cap = line.match(REGEX);
	places.add(cap[1]);
	places.add(cap[2]);
	distances[cap[1] + "-" + cap[2]] = Number(cap[3]);
	distances[cap[2] + "-" + cap[1]] = Number(cap[3]);
});
await new Promise(res => reader.once("close", res));

// --- Calculate distances ---
places = Array.from(places);
for (const p of permutations(places)) {
	let distance = 0;
	for (let i = 0; i < p.length - 1; i++) {
		distance += distances[p[i] + "-" + p[i + 1]];
	}

	min = Math.min(min, distance);
	max = Math.max(max, distance);
}

// --- Results ---
console.log("1. Min distance: ", min.toLocaleString());
console.log("2. Max distance: ", max.toLocaleString());

File: 2015/09/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func permutations(a []string) [][]string {
	var r [][]string

	var generate func(int)
	generate = func(i int) {
		if i == len(a) {
			tmp := make([]string, len(a))
			copy(tmp, a)
			r = append(r, tmp)
			return
		}
		for j := i; j < len(a); j++ {
			a[i], a[j] = a[j], a[i]
			generate(i + 1)
			a[i], a[j] = a[j], a[i]
		}
	}
	generate(0)
	return r
}

// --- Constants ---
const FILENAME = "../input/input.txt"
const REGEX = `^(.+) to (.+) = (\d+)$`

func main() {
	// --- Variables ---
	re := regexp.MustCompile(REGEX)
	set := make(map[string]bool)
	distances := make(map[string]int)
	mn := math.MaxInt
	mx := 0

	// --- Read input file ---
	file, err := os.Open(FILENAME)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// --- Parse each line ---
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		m := re.FindStringSubmatch(line)
		from := m[1]
		to := m[2]
		dist, err := strconv.Atoi(m[3])
		if err != nil {
			log.Panic(err)
		}

		set[from] = true
		set[to] = true
		distances[from+"-"+to] = dist
		distances[to+"-"+from] = dist
	}
	places := make([]string, 0, len(set))
	for p := range set {
		places = append(places, p)
	}

	// --- Calculate distances ---
	for _, p := range permutations(places) {
		d := 0
		for i := 0; i < len(p)-1; i++ {
			d += distances[p[i]+"-"+p[i+1]]
		}
		mn = min(mn, d)
		mx = max(mx, d)
	}

	// --- Results ---
	fmt.Println("1. Min distance:", mn)
	fmt.Println("2. Max distance:", mx)
}
